File: modules/objc/dispatch/ExtSdkWrapper.m
//
//  ExtSdkWrapper.m
//
//
//  Created by 杜洁鹏 on 2019/10/8.
//

#import "ExtSdkWrapper.h"
#import "ExtSdkDispatch.h"
#import "ExtSdkToJson.h"

#define easemob_dispatch_main_async_safe(block)                                \
    if ([NSThread isMainThread]) {                                             \
        block();                                                               \
    } else {                                                                   \
        dispatch_async(dispatch_get_main_queue(), block);                      \
    }

static NSString *const TAG = @"ExtSdkWrapper";
@implementation ExtSdkWrapper

- (void)onResult:(nonnull id<ExtSdkCallbackObjc>)result
    withMethodType:(nonnull NSString *)methodType
         withError:(nullable EMError *)error
        withParams:(nullable NSObject *)params {
    NSLog(@"%@: onResult: %@, %@, %@", TAG, methodType,
          error ? [error toJsonObject] : @"", params ? params : @"");
    if (nil == error) {
        NSMutableDictionary *data = [NSMutableDictionary dictionary];
        if (params) {
            data[methodType] = params;
        }
        [result onSuccess:data];
    } else {
        NSMutableDictionary *data = [NSMutableDictionary dictionary];
        data[@"error"] = [error toJsonObject];
        [result onFail:error.code withExtension:data];
    }
}

- (void)onReceive:(NSString *)methodType
       withParams:(nullable NSObject *)params {
    [[ExtSdkDispatch getInstance] onReceive:methodType withParams:params];
}

- (BOOL)checkMessageParams:(nonnull id<ExtSdkCallbackObjc>)result
            withMethodType:(nonnull NSString *)methodType
               withMessage:(nullable EMChatMessage *)message {
    if (message == nil) {
        [self onResult:result
            withMethodType:methodType
                 withError:[EMError errorWithDescription:
                                        @"The message does not exist."
                                                    code:1]
                withParams:nil];
        return YES;
    }
    return NO;
}

- (BOOL)getMessageParams:(nonnull id<ExtSdkCallbackObjc>)result
          withMethodType:(nonnull NSString *)methodType
             withMessage:(nullable EMChatMessage *)message {
    if (message == nil) {
        [self onResult:result
            withMethodType:methodType
                 withError:nil
                withParams:nil];
        return YES;
    }
    return NO;
}

- (void)mergeMessageBody:(EMMessageBody *)msgBody
       withDBMessageBody:(EMMessageBody *)dbMsgBody {
    if (msgBody.type == EMMessageBodyTypeText) {
        EMTextMessageBody *body = (EMTextMessageBody *)msgBody;
        EMTextMessageBody *dbBody = (EMTextMessageBody *)dbMsgBody;
        dbBody.targetLanguages = body.targetLanguages;
    } else if (msgBody.type == EMMessageBodyTypeImage) {
        EMImageMessageBody *body = (EMImageMessageBody *)msgBody;
        EMImageMessageBody *dbBody = (EMImageMessageBody *)dbMsgBody;
        dbBody.displayName = body.displayName;
        dbBody.localPath = body.localPath;
        dbBody.remotePath = body.remotePath;
        dbBody.secretKey = body.secretKey;
        dbBody.fileLength = body.fileLength;
        dbBody.downloadStatus = body.downloadStatus;
        dbBody.size = body.size;
        dbBody.compressionRatio = body.compressionRatio;
        dbBody.thumbnailDisplayName = body.thumbnailDisplayName;
        dbBody.thumbnailLocalPath = body.thumbnailLocalPath;
        dbBody.thumbnailRemotePath = body.thumbnailRemotePath;
        dbBody.thumbnailSecretKey = body.thumbnailSecretKey;
        dbBody.thumbnailSize = body.thumbnailSize;
        dbBody.thumbnailFileLength = body.thumbnailFileLength;
        dbBody.thumbnailDownloadStatus = body.thumbnailDownloadStatus;
    } else if (msgBody.type == EMMessageBodyTypeVideo) {
        EMVideoMessageBody *body = (EMVideoMessageBody *)msgBody;
        EMVideoMessageBody *dbBody = (EMVideoMessageBody *)dbMsgBody;
        dbBody.displayName = body.displayName;
        dbBody.localPath = body.localPath;
        dbBody.remotePath = body.remotePath;
        dbBody.secretKey = body.secretKey;
        dbBody.fileLength = body.fileLength;
        dbBody.downloadStatus = body.downloadStatus;
        dbBody.duration = body.duration;
        dbBody.thumbnailLocalPath = body.thumbnailLocalPath;
        dbBody.thumbnailRemotePath = body.thumbnailRemotePath;
        dbBody.thumbnailSecretKey = body.thumbnailSecretKey;
        dbBody.thumbnailSize = body.thumbnailSize;
        dbBody.thumbnailDownloadStatus = body.thumbnailDownloadStatus;
    } else if (msgBody.type == EMMessageBodyTypeLocation) {
        EMLocationMessageBody *body = (EMLocationMessageBody *)msgBody;
        EMLocationMessageBody *dbBody = (EMLocationMessageBody *)dbMsgBody;
        dbBody.latitude = body.latitude;
        dbBody.longitude = body.longitude;
        dbBody.address = body.address;
        dbBody.buildingName = body.buildingName;
    } else if (msgBody.type == EMMessageBodyTypeVoice) {
        EMVoiceMessageBody *body = (EMVoiceMessageBody *)msgBody;
        EMVoiceMessageBody *dbBody = (EMVoiceMessageBody *)dbMsgBody;
        dbBody.displayName = body.displayName;
        dbBody.localPath = body.localPath;
        dbBody.remotePath = body.remotePath;
        dbBody.secretKey = body.secretKey;
        dbBody.fileLength = body.fileLength;
        dbBody.downloadStatus = body.downloadStatus;
        dbBody.duration = body.duration;
    } else if (msgBody.type == EMMessageBodyTypeFile) {
        EMFileMessageBody *body = (EMFileMessageBody *)msgBody;
        EMFileMessageBody *dbBody = (EMFileMessageBody *)dbMsgBody;
        dbBody.displayName = body.displayName;
        dbBody.localPath = body.localPath;
        dbBody.remotePath = body.remotePath;
        dbBody.secretKey = body.secretKey;
        dbBody.fileLength = body.fileLength;
        dbBody.downloadStatus = body.downloadStatus;
    } else if (msgBody.type == EMMessageBodyTypeCmd) {
        EMCmdMessageBody *body = (EMCmdMessageBody *)msgBody;
        EMCmdMessageBody *dbBody = (EMCmdMessageBody *)dbMsgBody;
        dbBody.action = body.action;
        dbBody.isDeliverOnlineOnly = body.isDeliverOnlineOnly;
    } else if (msgBody.type == EMMessageBodyTypeCustom) {
        EMCustomMessageBody *body = (EMCustomMessageBody *)msgBody;
        EMCustomMessageBody *dbBody = (EMCustomMessageBody *)dbMsgBody;
        dbBody.event = body.event;
        dbBody.customExt = body.customExt;
    } else if (msgBody.type == EMMessageBodyTypeCombine) {
        EMCombineMessageBody *body = (EMCombineMessageBody *)msgBody;
        EMCombineMessageBody *dbBody = (EMCombineMessageBody *)dbMsgBody;
        dbBody.title = body.title;
        dbBody.summary = body.summary;
        dbBody.compatibleText = body.compatibleText;
    }
}

- (void)mergeMessage:(EMChatMessage *)msg withDBMessage:(EMChatMessage *)dbMsg {
    //    dbMsg.messageId = msg.messageId;
    //    dbMsg.conversationId = msg.conversationId;
    //    dbMsg.direction = msg.direction;
    //    dbMsg.from = msg.from;
    //    dbMsg.to = msg.to;
    //    dbMsg.timestamp = msg.timestamp;
    dbMsg.localTime = msg.localTime;
    //    dbMsg.chatType = msg.chatType;
    dbMsg.status = msg.status;
    //    dbMsg.isReadAcked = msg.isReadAcked;
    dbMsg.isChatThreadMessage = msg.isChatThreadMessage;
    dbMsg.isNeedGroupAck = msg.isNeedGroupAck;
    //    dbMsg.isDeliverAcked = msg.isDeliverAcked;
    dbMsg.isRead = msg.isRead;
    dbMsg.isListened = msg.isListened;
    dbMsg.ext = msg.ext;
    //    dbMsg.priority = msg.priority;
    dbMsg.deliverOnlineOnly = msg.deliverOnlineOnly;
    dbMsg.receiverList = msg.receiverList;
    [self mergeMessageBody:msg.body withDBMessageBody:dbMsg.body];
}

- (EMConversation *)getConversation:(NSDictionary *)param {
    NSString *convId = param[@"convId"];
    EMConversationType convType =
        [EMConversation typeFromInt:[param[@"convType"] intValue]];
    BOOL isChatThread =
        param[@"isChatThread"] ? [param[@"isChatThread"] boolValue] : NO;
    BOOL createIfNotExist =
        param[@"createIfNeed"] ? [param[@"createIfNeed"] boolValue] : YES;
    return [[EMClient.sharedClient chatManager] getConversation:convId
                                                           type:convType
                                               createIfNotExist:createIfNotExist
                                                       isThread:isChatThread];
}

- (EMConversation *)getConversationFromMessage:(EMChatMessage *)msg {
    BOOL createIfNotExist = msg.body.type != EMMessageBodyTypeCmd;
    return [[EMClient.sharedClient chatManager]
         getConversation:msg.conversationId
                    type:(EMConversationType)msg.chatType
        createIfNotExist:createIfNotExist
                isThread:msg.isChatThreadMessage];
}

@end

File: native_src/java/com/easemob/ext_sdk/dispatch/ExtSdkWrapper.java
package com.easemob.ext_sdk.dispatch;

import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.easemob.ext_sdk.common.ExtSdkCallback;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMCmdMessageBody;
import com.hyphenate.chat.EMCombineMessageBody;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMCustomMessageBody;
import com.hyphenate.chat.EMFileMessageBody;
import com.hyphenate.chat.EMImageMessageBody;
import com.hyphenate.chat.EMLocationMessageBody;
import com.hyphenate.chat.EMMessage;
import com.hyphenate.chat.EMMessageBody;
import com.hyphenate.chat.EMNormalFileMessageBody;
import com.hyphenate.chat.EMTextMessageBody;
import com.hyphenate.chat.EMVideoMessageBody;
import com.hyphenate.chat.EMVoiceMessageBody;
import com.hyphenate.exceptions.HyphenateException;
import java.util.HashMap;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class ExtSdkWrapper {

    protected static void onSuccess(@NonNull ExtSdkCallback callback, @NonNull String methodType,
                                    @Nullable Object object) {
        Log.d(TAG, "onSuccess: " + methodType + ": " + (object != null ? object : ""));
        Map<String, Object> data = new HashMap<>();
        if (object != null) {
            data.put(methodType, object);
        }
        callback.success(data);
    }

    public static void onError(@NonNull ExtSdkCallback callback, @NonNull Object e, @Nullable Object ext) {
        Log.d(TAG, "onError: " + e + ": " + (ext != null ? ext : ""));
        Map<String, Object> data = new HashMap<>();
        if (e instanceof HyphenateException) {
            data.put("error", ExtSdkExceptionHelper.toJson((HyphenateException)e));
        } else if (e instanceof Integer) {
            data.put("error", ExtSdkErrorHelper.toJson((int)e, ext != null ? ext.toString() : ""));
        } else if (e instanceof JSONException) {
            data.put("error", ExtSdkJSONExceptionHelper.toJson((JSONException)e));
        } else {
            data.put("error", "no implement");
        }
        callback.success(data);
    }

    protected static void onReceive(@NonNull String methodType, @Nullable Object data) {
        Log.d(TAG, "onReceive: " + methodType + ": " + (data != null ? data : ""));
        ExtSdkDispatch.getInstance().onReceive(methodType, data);
    }

    protected void mergeMessageBody(EMMessageBody msgBody, EMMessage dbMsg) throws JSONException {
        if (dbMsg.getType() == EMMessage.Type.TXT) {
            EMTextMessageBody text = (EMTextMessageBody)msgBody;
            EMTextMessageBody dbtext = (EMTextMessageBody)dbMsg.getBody();
            dbtext.setMessage(text.getMessage());
            dbtext.setTargetLanguages(text.getTargetLanguages());
            dbMsg.setBody(dbtext);
        } else if (dbMsg.getType() == EMMessage.Type.CMD) {
            EMCmdMessageBody cmd = (EMCmdMessageBody)msgBody;
            EMCmdMessageBody dbcmd = (EMCmdMessageBody)dbMsg.getBody();
            dbcmd.deliverOnlineOnly(cmd.isDeliverOnlineOnly());
            dbMsg.setBody(dbcmd);
        } else if (dbMsg.getType() == EMMessage.Type.IMAGE) {
            EMImageMessageBody image = (EMImageMessageBody)msgBody;
            EMImageMessageBody dbimage = (EMImageMessageBody)dbMsg.getBody();
            dbimage.setFileName(image.getFileName());
            dbimage.setLocalUrl(image.getLocalUrl());
            dbimage.setRemoteUrl(image.getRemoteUrl());
            dbimage.setSecret(image.getSecret());
            dbimage.setFileLength(image.getFileSize());
            dbimage.setDownloadStatus(image.downloadStatus());
            dbimage.setSendOriginalImage(image.isSendOriginalImage());
            dbimage.setThumbnailLocalPath(image.thumbnailLocalPath());
            dbimage.setThumbnailDownloadStatus(image.thumbnailDownloadStatus());
            dbimage.setThumbnailUrl(image.getThumbnailUrl());
            //        dbimage.setThumbnailSize(image.getWidth(), image.getHeight());
            dbMsg.setBody(dbimage);
        } else if (dbMsg.getType() == EMMessage.Type.VOICE) {
            EMVoiceMessageBody voice = (EMVoiceMessageBody)msgBody;
            EMVoiceMessageBody dbvoice = (EMVoiceMessageBody)dbMsg.getBody();
            dbvoice.setFileName(voice.getFileName());
            dbvoice.setLocalUrl(voice.getLocalUrl());
            dbvoice.setRemoteUrl(voice.getRemoteUrl());
            dbvoice.setSecret(voice.getSecret());
            dbvoice.setFileLength(voice.getFileSize());
            dbvoice.setDownloadStatus(voice.downloadStatus());
            dbMsg.setBody(dbvoice);
        } else if (dbMsg.getType() == EMMessage.Type.VIDEO) {
            EMVideoMessageBody video = (EMVideoMessageBody)msgBody;
            EMVideoMessageBody dbvideo = (EMVideoMessageBody)dbMsg.getBody();
            dbvideo.setFileName(video.getFileName());
            dbvideo.setLocalUrl(video.getLocalUrl());
            dbvideo.setRemoteUrl(video.getRemoteUrl());
            dbvideo.setSecret(video.getSecret());
            dbvideo.setDownloadStatus(video.downloadStatus());
            dbvideo.setVideoFileLength(video.getVideoFileLength());
            dbvideo.setThumbnailUrl(video.getThumbnailUrl());
            dbvideo.setThumbnailSize(video.getThumbnailWidth(), video.getThumbnailHeight());
            dbvideo.setLocalThumb(video.getLocalThumb());
            dbvideo.setThumbnailSecret(video.getThumbnailSecret());
            dbvideo.setThumbnailDownloadStatus(video.thumbnailDownloadStatus());
            dbMsg.setBody(video);
        } else if (dbMsg.getType() == EMMessage.Type.LOCATION) {
            EMLocationMessageBody location = (EMLocationMessageBody)msgBody;
            dbMsg.setBody(location);
        } else if (dbMsg.getType() == EMMessage.Type.FILE) {
            EMNormalFileMessageBody file = (EMNormalFileMessageBody)msgBody;
            EMFileMessageBody dbfile = (EMFileMessageBody)dbMsg.getBody();
            dbfile.setFileName(file.getFileName());
            dbfile.setLocalUrl(file.getLocalUrl());
            dbfile.setRemoteUrl(file.getRemoteUrl());
            dbfile.setSecret(file.getSecret());
            dbfile.setDownloadStatus(file.downloadStatus());
            dbfile.setFileLength(file.getFileSize());
            dbMsg.setBody(dbfile);
        } else if (dbMsg.getType() == EMMessage.Type.CUSTOM) {
            EMCustomMessageBody cutom = (EMCustomMessageBody)msgBody;
            dbMsg.setBody(cutom);
        } else if (dbMsg.getType() == EMMessage.Type.COMBINE) {
            EMCombineMessageBody combine = (EMCombineMessageBody)msgBody;
            dbMsg.setBody(combine);
        }
    }

    protected void mergeMessage(EMMessage msg, EMMessage dbMsg) throws JSONException {
        dbMsg.setStatus(msg.status());
        //      dbMsg.setMsgTime(msg.getMsgTime());
        dbMsg.setLocalTime(msg.localTime());
        dbMsg.setIsNeedGroupAck(msg.isNeedGroupAck());
        //      dbMsg.setGroupAckCount(msg.groupAckCount());
        dbMsg.setIsChatThreadMessage(msg.isChatThreadMessage());
        //      dbMsg.setFrom(msg.getFrom());
        //      dbMsg.setTo(msg.getTo());
        //      dbMsg.setMsgId(msg.getMsgId());
        //      dbMsg.setChatType(msg.getChatType());
        //      dbMsg.setAcked(msg.isAcked());
        //      dbMsg.setDelivered(msg.isDelivered());
        dbMsg.setUnread(msg.isUnread());
        dbMsg.setListened(msg.isListened());
        //      dbMsg.setDirection(msg.direct());
        dbMsg.setReceiverList(msg.receiverList());
        //      dbMsg.setPriority();
        Map<String, Object> list = msg.getAttributes();
        if (list.size() > 0) {
            JSONObject jsonParams = new JSONObject(list);
            for (Map.Entry<String, Object> entry : list.entrySet()) {
                String key = entry.getKey();
                Object result = entry.getValue();
                if (result.getClass().getSimpleName().equals("Integer")) {
                    dbMsg.setAttribute(key, (Integer)result);
                } else if (result.getClass().getSimpleName().equals("Boolean")) {
                    dbMsg.setAttribute(key, (Boolean)result);
                } else if (result.getClass().getSimpleName().equals("Long")) {
                    dbMsg.setAttribute(key, (Long)result);
                } else if (result.getClass().getSimpleName().equals("Double") ||
                           result.getClass().getSimpleName().equals("Float")) {
                    dbMsg.setAttribute(key, (Double)result);
                } else if (result.getClass().getSimpleName().equals("JSONObject")) {
                    dbMsg.setAttribute(key, (JSONObject)result);
                } else if (result.getClass().getSimpleName().equals("JSONArray")) {
                    dbMsg.setAttribute(key, (JSONArray)result);
                } else {
                    dbMsg.setAttribute(key, jsonParams.getString(key));
                }
            }
        }

        this.mergeMessageBody(msg.getBody(), dbMsg);
    }

    protected EMConversation getConversation(JSONObject params) throws JSONException {
        String convId = params.getString("convId");
        EMConversation.EMConversationType convType = ExtSdkConversationHelper.typeFromInt(params.getInt("convType"));
        boolean isChatThread = params.optBoolean("isChatThread", false);
        boolean createIfNotExists = params.optBoolean("createIfNeed", true);
        return EMClient.getInstance().chatManager().getConversation(convId, convType, createIfNotExists, isChatThread);
    }

    protected EMConversation getConversationFromMessage(EMMessage message) {
        boolean createConv = message.getType() != EMMessage.Type.CMD;
        return EMClient.getInstance().chatManager().getConversation(
            message.conversationId(), EMConversation.msgType2ConversationType(message.getTo(), message.getChatType()),
            createConv, message.isChatThreadMessage());
    }

    protected static boolean checkMessageParams(EMMessage msg, String channelName, ExtSdkCallback result)
        throws JSONException {
        if (msg == null) {
            ExtSdkWrapper.onError(result, 1, "The message does not exist.");
            return true;
        }
        return false;
    }

    protected static boolean getMessageParams(EMMessage msg, String channelName, ExtSdkCallback result)
        throws JSONException {
        if (msg == null) {
            ExtSdkWrapper.onSuccess(result, channelName, null);
            return true;
        }
        return false;
    }

    private static final String TAG = "EMWrapper";
}
